lib/wasi/src/state/builder.rs

```diff
@@ -41,16 +40,20 @@ pub(crate) fn create_wasi_state(program_name: &str) -> WasiStateBuilder {
 /// ```
 #[derive(Default)]
 pub struct WasiStateBuilder {
-    args: Vec<Vec<u8>>,
-    envs: Vec<(Vec<u8>, Vec<u8>)>,
+    args: Vec<String>,
+    envs: Vec<(String, Vec<u8>)>,
     preopens: Vec<PreopenedDir>,
+    uses: Vec<String>,
+    #[cfg(feature = "sys")]
+    map_commands: HashMap<String, PathBuf>,
     vfs_preopens: Vec<String>,
+    compiled_modules: Arc<ModuleCache>,
     #[allow(clippy::type_complexity)]
     setup_fs_fn: Option<Box<dyn Fn(&mut WasiInodes, &mut WasiFs) -> Result<(), String> + Send>>,
     stdout_override: Option<Box<dyn VirtualFile + Send + Sync + 'static>>,
     stderr_override: Option<Box<dyn VirtualFile + Send + Sync + 'static>>,
     stdin_override: Option<Box<dyn VirtualFile + Send + Sync + 'static>>,
-    fs_override: Option<Box<dyn wasmer_vfs::FileSystem>>,
+    fs_override: Option<WasiFsRoot>,
```

Comment:
It shouldn't be a root? Perhaps I'm missing the context on `WasiFsRoot`. Can someone shed some light here?

Reply:
`WasiFsRoot` is just a simple enum to distinguish between a dynamic backing filesystem and a concrete `TmpFs`. 

I bet this materialized because John needed it to union the file system with others for merging WebC volumes. 

I agree that this should just be stripped out.

We have a few options of how to go about this.

* always use a TmpFs as the root, and "mount" other file systems in the TmpFs as needed
* Always use an `Arc<dyn FileSystem>`   (note  that it needs to be arc now with thread spawning etc)
    In that case we probably want to add an `as_any(&self) -> Option<&dyn Any>` method to the trait to enable downcasting
   
Note: I'm also removing the `.uses()` method from the WasiEnv builder and moving all the webc related setup code to a single place (BinFactory), so we can initially just use `dyn`.
But we might want to dynamically union in new filesystems in the future for Deploy, just to put this on the radar.

Reply:
Asked and answered